controllers/workspace/predicates.go

```diff
@@ -0,0 +1,41 @@
+//
+// Copyright (c) 2019-2020 Red Hat, Inc.
+// This program and the accompanying materials are made
+// available under the terms of the Eclipse Public License 2.0
+// which is available at https://www.eclipse.org/legal/epl-2.0/
+//
+// SPDX-License-Identifier: EPL-2.0
+//
+// Contributors:
+//   Red Hat, Inc. - initial API and implementation
+//
+
+package controllers
+
+import (
+	"github.com/devfile/api/pkg/apis/workspaces/v1alpha2"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+)
+
+var predicates = predicate.Funcs{
+	CreateFunc: func(_ event.CreateEvent) bool { return true },
+	DeleteFunc: func(_ event.DeleteEvent) bool { return true },
+	UpdateFunc: func(ev event.UpdateEvent) bool {
+		newObj, ok := ev.ObjectNew.(*v1alpha2.DevWorkspace)
+		if !ok {
+			return true
+		}
+		if newObj.Status.Phase != v1alpha2.WorkspaceStatusFailed {
+			return true
+		}
+		oldObj, ok := ev.ObjectOld.(*v1alpha2.DevWorkspace)
+		if !ok {
+			// Should never happen
+			return true
+		}
+		// Trigger a reconcile on failed workspaces if routingClass is updated.
+		return oldObj.Spec.RoutingClass != newObj.Spec.RoutingClass
```

Comment:
I see that it's useful to reconcile workspace if any changes made in spec, like the following scenario:

```bash
# 1. Create DevWorkspace
cat <<EOF | kubectl apply -f - 
kind: DevWorkspace
apiVersion: workspace.devfile.io/v1alpha2
metadata:
  name: theia
spec:
  started: true
  template:
    projects:
      - name: project
        git:
          remotes:
            origin: "https://github.com/che-samples/web-nodejs-sample.git"
    components:
      - name: theia
        plugin:
          id: eclipse/che-theia/next
      - name: terminal
        plugin:
          id: eclipse/che-machine-exec-plugin/non-existing
EOF
# 2.
kubectl get dw theia  # make sure it's failing
# 3.
kubectl patch dw theia --type merge --patch '
spec:
  template:
    components:
      - name: theia
        plugin:
          id: eclipse/che-theia/next
      - name: terminal
        plugin:
          id: eclipse/che-machine-exec-plugin/nightly
'
# 4. DevWorkspace could be started correctly now but it's still failing
kc get dw theia
```

Reply:
Fixed by c18edb9